server/src/main/java/org/elasticsearch/search/aggregations/support/BuiltinValuesSourceType.java

```diff
@@ -0,0 +1,214 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.search.aggregations.support;
+
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.common.geo.GeoPoint;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.index.fielddata.IndexFieldData;
+import org.elasticsearch.index.fielddata.IndexGeoPointFieldData;
+import org.elasticsearch.index.fielddata.IndexNumericFieldData;
+import org.elasticsearch.index.fielddata.IndexOrdinalsFieldData;
+import org.elasticsearch.index.mapper.MappedFieldType;
+import org.elasticsearch.index.mapper.RangeFieldMapper;
+import org.elasticsearch.script.AggregationScript;
+import org.elasticsearch.search.DocValueFormat;
+import org.elasticsearch.search.aggregations.AggregationExecutionException;
+
+import java.io.IOException;
+import java.util.Locale;
+import java.util.function.LongSupplier;
+
+public enum BuiltinValuesSourceType implements Writeable, ValuesSourceType {
+    ANY {
+        @Override
+        public ValuesSource getEmpty() {
+            // TODO: Implement this or get rid of ANY
+            throw new UnsupportedOperationException("ValuesSourceType.ANY is still a special case");
+        }
+
+        @Override
+        public ValuesSource getScript(AggregationScript.LeafFactory script, ValueType scriptValueType) {
+            // TODO: Implement this or get rid of ANY
+            throw new UnsupportedOperationException("ValuesSourceType.ANY is still a special case");
+        }
+
+        @Override
+        public ValuesSource getField(FieldContext fieldContext, AggregationScript.LeafFactory script) {
+            // TODO: Implement this or get rid of ANY
+            throw new UnsupportedOperationException("ValuesSourceType.ANY is still a special case");
+        }
+
+        @Override
+        public ValuesSource replaceMissing(ValuesSource valuesSource, Object rawMissing, DocValueFormat docValueFormat, LongSupplier now) {
+            return BYTES.replaceMissing(valuesSource, rawMissing, docValueFormat, now);
+        }
+    },
+    NUMERIC {
+        @Override
+        public ValuesSource getEmpty() {
+            return ValuesSource.Numeric.EMPTY;
+        }
+
+        @Override
+        public ValuesSource getScript(AggregationScript.LeafFactory script, ValueType scriptValueType) {
+            return new ValuesSource.Numeric.Script(script, scriptValueType);
+        }
+
+        @Override
+        public ValuesSource getField(FieldContext fieldContext, AggregationScript.LeafFactory script) {
+
+            if (!(fieldContext.indexFieldData() instanceof IndexNumericFieldData)) {
+                // TODO: Is this the correct exception type here?
+                throw new IllegalArgumentException("Expected numeric type on field [" + fieldContext.field() +
```

Comment:
Hmmm.... maybe IllegalState instead?

Reply:
So, I copied the exception types from the old behavior as much as possible.  I do think it's confusing what exceptions get thrown from where, but I have BWC concerns with changing exception types, and think we need a bigger effort to standardize what we throw when from aggs.  Given that, I'd rather leave this as is for now.